packages/next/src/build/webpack/plugins/slow-module-detection-plugin.ts

```diff
@@ -0,0 +1,330 @@
+    const rootModules = [...this.moduleNodes.values()].filter(
+      (node) => !this.moduleParents.has(node)
+    )
+
+    // Helper to truncate long paths for better readability
+    const truncatePath = (
+      path: string,
+      maxLength = this.pathTruncationLength
+    ): string => {
+      if (path.length <= maxLength) return path
+      const startSegment = path.slice(0, maxLength / 3)
+      const endSegment = path.slice((-maxLength * 2) / 3)
+      return `${startSegment}...${endSegment}`
+    }
+
+    // Helper to format duration with consistent padding
+    const formatDuration = (ms: number): string => {
+      return `${ms}ms`.padStart(5)
+    }
+
+    // Recursively format a module and its children
+    const formatModuleNode = (node: ModuleNode, depth: number): string => {
+      const moduleName = getModuleDisplayName(node.module) || ''
+
+      if (!moduleName) {
+        return formatChildModules(node, depth)
+      }
+
+      const truncatedName = truncatePath(moduleName)
+      const indentation = TreeSymbols.VERTICAL_LINE.repeat(depth)
+
+      const duration =
+        node.buildDuration > 0
+          ? `(${ConsoleColors.BRIGHT_YELLOW}${formatDuration(node.buildDuration)}${ConsoleColors.RESET})`
+          : ''
+      const nameWithStyle = `${ConsoleColors.BOLD_BLUE}${truncatedName}${ConsoleColors.RESET}`
+      const line = `${indentation}${TreeSymbols.BRANCH}${nameWithStyle} ${duration}`
+
+      return ` ${line}\n${formatChildModules(node, depth + 1)}`
+    }
+
+    // Format all children of a module
+    const formatChildModules = (node: ModuleNode, depth: number): string => {
+      const children = this.moduleChildren.get(node)
+      if (!children) return ''
+
+      return [...children]
+        .map(([_, child]) => formatModuleNode(child, depth))
+        .join('')
+    }
+
+    // Build the complete report
+    let report = ''
+    for (const root of rootModules) {
+      report += formatModuleNode(root, 0)
+    }
+
+    // Only output if there are slow modules to report
+    if (report) {
+      console.log(
+        `${ConsoleColors.BOLD_GREEN}🐌 Detected slow modules while compiling ${this.options.compilerType}:${ConsoleColors.RESET}`
+      )
+      console.log(report)
+    }
+  }
+}
+
+/**
+ * Webpack plugin that analyzes and reports modules that are slow to build.
+ * Hooks into webpack's compilation process to:
+ * 1. Track when module builds start
+ * 2. Record build durations when modules complete
+ * 3. Generate a report when compilation is done
+ */
+export default class SlowModuleDetectionPlugin {
+  private analyzer: ModuleBuildTimeAnalyzer | null = null
+  private moduleBuildStartTimes = new WeakMap<Module, number>()
+
+  constructor(private options: ModuleBuildTimeAnalyzerOptions) {}
+
+  apply = (compiler: Compiler) => {
+    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
+      this.analyzer = new ModuleBuildTimeAnalyzer(this.options)
```

Comment:
You'll want to keep these on a WeakMap keyed by `compilation` given that there could be multiple at the same time.